### DISCUSSION TITLE
Show data written offline when using offlineExchange?

### DISCUSSION THREAD

**Discussion by therealjbenam** [urql-graphql/urql#2952] [Q&A] [answered]
Hi,

not sure if this is a bug or it's supposed to work like this, but I have set up an offlineExchange like this (following https://formidable.com/open-source/urql/docs/graphcache/offline/ ) with a MongoDB backend:

```
// Client-side-only code

let storage;

if (typeof window !== 'undefined') {
  storage = makeDefaultStorage({
    idbName: 'graphcache-v3',
    maxAge: 7,
  });

const cache = offlineExchange({
  schema: introspectedSchema,
  storage,
  updates: {},
  optimistic: {
    insertOneEntity(variables, cache, info) {
      return {
        __typename: 'Entity',
        id: variables._id,
        test: 'sample',
      };
    },
  },
});

const client = createClient({
  url: `<snip>`,
  exchanges: [dedupExchange, cache, fetchExchange],
  fetchOptions: () => {
    const accessToken = authGetToken();
    return {
      headers: { Authorization: `Bearer ${accessToken}` },
    };
  },
```

This is how I am retrieving the data:

```
  const EntitiesListQuery = gql`
    query {
      entity(limit: 0) {
        _id
        test
      }
    }
  `;

useQuery({
    query: EntityListQuery,
    requestPolicy: 'cache-and-network',
  });
```

And this is my mutation:
```
  const testMutation = gql`
    mutation {
      insertOneEntity(data: { test: "sample" }) {
        _id
      }
    }
  `;

  const [result, executeMutation] = useMutation(testMutation);
```

It works fine in offline mode when reading cached data - my issue is when writing something new.

I can see the data being added to `metadata`, but it doesn't show up in the UI. Going back online allows the data to sync and it appears normally, but I would like to display it offline beforehand to the user, like Firebase does - allowing a true offline experience.

From what I can read in the docs, it says "we can set up [Optimistic Updates] for these mutations, which allows them to be reflected in our UI before sending a request to the API." - as such it should be working kinda like Firebase, but I can't get it to work like that.

I have also tried implementing a custom storage using the examples provided, which should apparently send "complete data" but no dice.

Not sure if it's a case of the docs being unclear (and this is intended behaviour) or if I am doing something wrong.

Can someone please shed some light on this?

BTW, thank you very much for the great job you have done with urql.

**Comment by JoviDeCroock** [comment, marked as answer]
When you are inserting entities you need to write an updater to insert into the list, we can't assume that it belongs in list x. This is described in the [docs](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#updating-lists-or-links).

We can automatically update existing entities but we can't assume that `create`/`delete` links or unlinks from i.e. Query.entities

**Comment by therealjbenam** [comment]
> When you are inserting entities you need to write an updater to insert into the list, we can't assume that it belongs in list x. This is described in the [docs](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#updating-lists-or-links).
> 
> We can automatically update existing entities but we can't assume that `create`/`delete` links or unlinks from i.e. Query.entities

I understand, thanks. I have tried adding the following - not sure if it's correct - but I still don't get any updates when offline.

```
const cache = offlineExchange({
  schema: introspectedSchema,
  storage,
  updates: {
    Mutation: {
      insertOneEntity(result, _args, cache, _info) {
        const EntityList = gql`
          mutation {
            insertOneEntity(data: { test: 'sample' }) {
              _id
            }
          }
        `;

        cache.updateQuery({ query: EntityList }, data => {
          data.entities.push(result.insertOneEntity);
          return data;
        });
      },
    },
  },
  optimistic: {
    insertOneEntity(variables, cache, info) {
      return {
        __typename: 'Entity',
        id: variables._id,
        test: 'sample',
      };
    },
  },
});
```

**Comment by JoviDeCroock** [comment]
You are updating a query but providing a mutation 😅 you would need to do 

```
        cache.updateQuery({ query: gql`query { entities { _id text ... }}` }, data => {
          data.entities.push(result.insertOneEntity);
          return data;
        });
```

or use the `cache.link` helper
